screwdriver/screwdriver.go

```diff
@@ -0,0 +1,183 @@
+		UserToken:  token,
+	}
+
+	jwt, err := s.jwt()
+	if err != nil {
+		return s, err
+	}
+
+	s.SDJWT = jwt
+
+	return s, nil
+}
+
+func (sd *sdAPI) makeURL(path string) (*url.URL, error) {
+	version := "v4"
+	fullpath := fmt.Sprintf("%s/%s/%s", sd.APIURL, version, path)
+
+	return url.Parse(fullpath)
+}
+
+func (sd *sdAPI) request(method, path string, body io.Reader) (*http.Response, error) {
+	req, err := http.NewRequest(method, path, body)
+	if err != nil {
+		return nil, err
+	}
+
+	switch method {
+	case http.MethodGet:
+		{
+			req.Header.Add("Accept", "application/json")
+		}
+	case http.MethodPost, http.MethodPut, http.MethodDelete:
+		{
+			req.Header.Add("Content-Type", "application/json")
+			req.Header.Add("Authorization", "Bearer "+sd.SDJWT)
+		}
+	}
+
+	return sd.HTTPClient.Do(req)
+}
+
+func (sd *sdAPI) jwt() (string, error) {
+	path := "auth/token?api_token=" + sd.UserToken
+	fullpath, err := sd.makeURL(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to send request: %v", err)
+	}
+
+	res, err := sd.request(http.MethodGet, fullpath.String(), nil)
+	if err != nil {
+		return "", err
+	}
+	if res.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("failed to get JWT: StatusCode %d", res.StatusCode)
+	}
+	defer res.Body.Close()
+
+	tokenResponse := new(tokenResponse)
+	err = json.NewDecoder(res.Body).Decode(tokenResponse)
+
+	return tokenResponse.JWT, err
+}
+
+func readScrewdriverYAML(filePath string) (string, error) {
+	yaml, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read screwdriver.yaml: %v", err)
+	}
+	return string(yaml), nil
+}
+
+func (sd *sdAPI) validate(filePath string) (*validatorResponse, error) {
+	fullpath, err := sd.makeURL("validator")
+	if err != nil {
+		return &validatorResponse{}, fmt.Errorf("failed to create api endpoint URL: %v", err)
+	}
+
+	yaml, err := readScrewdriverYAML(filePath)
+	if err != nil {
+		return &validatorResponse{}, err
+	}
+
+	escapedYaml := strings.ReplaceAll(string(yaml), "\"", "\\\"")
+	body := fmt.Sprintf(`{"yaml": "%s"}`, strings.ReplaceAll(string(escapedYaml), "\n", "\\n"))
+
+	res, err := sd.request(http.MethodPost, fullpath.String(), bytes.NewBuffer([]byte(body)))
+	if err != nil {
+		return &validatorResponse{}, fmt.Errorf("failed to send request: %v", err)
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return &validatorResponse{}, fmt.Errorf("failed to post validator: StatusCode %d", res.StatusCode)
+	}
+
+	v := new(validatorResponse)
+	err = json.NewDecoder(res.Body).Decode(v)
+	if err != nil {
+		return &validatorResponse{}, fmt.Errorf("failed to parse validator response: %v", err)
+	}
+
+	return v, nil
+}
+
+// Job returns job represented by "jobName"
+func (sd *sdAPI) Job(jobName, filepath string) (Job, error) {
+	v, err := sd.validate(filepath)
+	if err != nil {
+		return Job{}, err
+	}
+
+	if v.Errors != nil {
+		return Job{}, fmt.Errorf("failed to parse screwdriver.yaml: %v", v.Errors)
+	}
+
+	job, ok := v.Jobs[jobName]
```

Comment:
I wonder why does this return slice?

Reply:
`/v4/validator` endpoint returns JSON like below.

```json
{
  "annotations": {},
  "jobs": {
    "main": [
      {
        "annotations": {},
        "commands": [
          {
            "name": "start",
            "command": "echo start"
          },
          {
            "name": "end",
            "command": "echo end"
          }
        ],
        "environment": {
          "HOGE": "1"
        },
        "image": "alpine",
        "secrets": [],
        "settings": {}
      }
    ],
    "test": [
      {
        "annotations": {},
        "commands": [
          {
            "name": "test",
            "command": "echo test"
          }
        ],
        "environment": {},
        "image": "alpine",
        "secrets": [],
        "settings": {}
      }
    ]
  },
  "workflowGraph": {
    "nodes": [
      {
        "name": "~pr"
      },
      {
        "name": "~commit"
      },
      {
        "name": "main"
      },
      {
        "name": "test"
      }
    ],
    "edges": []
  },
  "parameters": {}
}
```

`jobs` is key-value pair and keys are the job name and the value are an array consist of the job for it. 
So, it is strange but we should take the first value of this slice.

Reply:
I got it 😥